src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -225,12 +204,113 @@ impl StreamingUploader for OpenDalStreamingUploader {
     }
 }
 
+struct OpenDalObjectIter {
+    lister: Option<Lister>,
+    op: Option<Operator>,
+    #[allow(clippy::type_complexity)]
+    next_future: Option<BoxFuture<'static, (Option<Result<Entry, Error>>, Lister)>>,
+    #[allow(clippy::type_complexity)]
+    metadata_future: Option<BoxFuture<'static, (Result<ObjectMetadata, Error>, Operator)>>,
+}
+
+impl OpenDalObjectIter {
+    fn new(lister: Lister, op: Operator) -> Self {
+        Self {
+            lister: Some(lister),
+            op: Some(op),
+            next_future: None,
+            metadata_future: None,
+        }
+    }
+}
+
+impl Stream for OpenDalObjectIter {
+    type Item = ObjectResult<ObjectMetadata>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if let Some(metadata_future) = self.metadata_future.as_mut() {
+            let (result, op) = ready!(metadata_future.poll_unpin(cx));
+            self.op = Some(op);
+            return match result {
+                Ok(m) => {
+                    self.metadata_future = None;
+                    Poll::Ready(Some(Ok(m)))
+                }
+                Err(e) => {
+                    self.metadata_future = None;
+                    Poll::Ready(Some(Err(e.into())))
+                }
+            };
+        }
+        if let Some(next_future) = self.next_future.as_mut() {
+            let (option, lister) = ready!(next_future.poll_unpin(cx));
+            self.lister = Some(lister);
+            return match option {
+                None => {
+                    self.next_future = None;
+                    Poll::Ready(None)
+                }
+                Some(result) => {
+                    self.next_future = None;
+                    match result {
+                        Ok(object) => {
+                            let op = self.op.take().expect("op should not be None");
+                            let f = async move {
+                                let key = object.path().to_string();
+                                // FIXME: How does opendal metadata cache work?
+                                // Will below line result in one IO per object?
```

Comment:
I haven't confirmed this in the PR. Any idea?
@wcy-fdu

Reply:
Let‘s verify this later.

Reply:
Generally LGTM for this part, since we don't have test pipeline on gcs or other opendal-supported storage, I think we can merge it first, and verify it via daily test on gcp(cc @huangjw806).

Could you please take a look if you have time? cc @Xuanwo

Reply:
Any update?

Reply:
The new metakey API list will be released in opendal v0.40. The current implementation is efficient and does not require additional calls for most storage services.